test/unit/hpa.c

```diff
@@ -452,6 +467,36 @@ TEST_BEGIN(test_defer_time) {
 }
 TEST_END
 
+TEST_BEGIN(test_purge_no_infinite_loop) {
+	test_skip_if(!hpa_supported());
+
+	hpa_shard_t *shard = create_test_data(&hpa_hooks_default,
+	    &test_hpa_shard_opts_purge);
+	tsdn_t *tsdn = tsd_tsdn(tsd_fetch());
+
+	/*
+	 * This is not arbitrary value, it is chosen to met hugification
+	 * criteria for huge page and at the same time do not allow hugify page
+	 * without triggering a purge.
+	 */
+	const size_t npages =
+	    test_hpa_shard_opts_purge.hugification_threshold / PAGE + 1;
+	const size_t size = npages * PAGE;
+
+	bool deferred_work_generated = false;
+	edata_t *edata = pai_alloc(tsdn, &shard->pai, size, PAGE,
+	     /* zero */ false, /* guarded */ false, /* frequent_reuse */ false,
+	     &deferred_work_generated);
+	expect_ptr_not_null(edata, "Unexpected alloc failure");
+
+	hpa_shard_do_deferred_work(tsdn, shard);
```

Comment:
Cool neat reproduction! It seems the failure rate is 100% now without the fix, right?

Reply:
> It seems the failure rate is 100% now without the fix, right?

Yes, it triggers 100% at least on x86_64, with `PAGE = 4 KiB` and `HUGEPAGE = 2 MiB`.

Basically, we allocate 461 page out of 512 pages on the huge page, which is 1888256 bytes (461 * 4 KiB) with 1887436 bytes as hugification threshold (2 MiB * 0.9). So, `hpa_should_purge` returns `true`, because `hpa_hugify_blocked_by_ndirty` returns `true`. We enter purging loop and call `hpa_try_purge` to purge, but there is nothing to purge, so it returns `false`, we didn't increment `nops` and continue `while` from the beginning again.